model/metadata/system_test.go

```diff
@@ -18,165 +18,181 @@
 package metadata
 
 import (
+	"encoding/json"
 	"errors"
+	"fmt"
+	"strings"
 	"testing"
 
+	"github.com/stretchr/testify/require"
+
 	"github.com/stretchr/testify/assert"
 
+	"github.com/elastic/apm-server/tests/approvals"
 	"github.com/elastic/apm-server/utility"
-	"github.com/elastic/beats/libbeat/common"
 )
 
-func TestSystemTransform(t *testing.T) {
-
-	architecture := "x64"
-	hostname, name := "a.b.com", "foo"
-	platform := "darwin"
-	ip := "127.0.0.1"
+func TestSystem(t *testing.T) {
+	host, configured, detected := "host", "custom hostname", "detected hostname"
+	arch, platform, ip, containerID, namespace := "amd", "osx", "127.0.0.1", "1234", "staging"
 	empty := ""
-	nodename := "a.node"
-	podname := "a.pod"
+	nodename, podname, podUID := "a.node", "a.pod", "b.podID"
 
-	tests := []struct {
-		System System
-		Output common.MapStr
+	inpErr := errors.New("some error")
+	for name, test := range map[string]struct {
+		input         interface{}
+		inputErr, err error
+		s             *System
 	}{
-		{
-			System: System{},
-			Output: common.MapStr{},
+		"nil":         {input: nil, err: nil, s: nil},
+		"inputError":  {input: nil, inputErr: inpErr, err: inpErr, s: nil},
+		"invalidType": {input: "", err: errors.New("invalid type for system"), s: nil},
+		"empty":       {input: map[string]interface{}{}, s: &System{}},
+		"fetchError": {
+			input: map[string]interface{}{"hostname": 1},
+			err:   utility.ErrFetch,
+			s:     &System{DetectedHostname: nil, Architecture: nil, Platform: nil, IP: nil},
 		},
-		{
-			System: System{
-				IP: &empty,
-			},
-			Output: common.MapStr{},
-		},
-		{
-			System: System{
-				Architecture: &architecture,
-				Hostname:     &hostname,
-				Name:         &name,
-				Platform:     &platform,
-				IP:           &ip,
-			},
-			Output: common.MapStr{
-				"architecture": architecture,
-				"hostname":     hostname,
-				"name":         name,
-				"ip":           ip,
-				"os": common.MapStr{
-					"platform": platform,
-				},
+		"empty ip": {
+			input: map[string]interface{}{"ip": ""},
+			s:     &System{IP: &empty},
+		},
+		"hostname": {
+			input: map[string]interface{}{"hostname": host},
+			s:     &System{DetectedHostname: &host},
+		},
+		"detected hostname": {
```

Comment:
This shouldn't happen in practice right?  I like testing the path, just want to make sure I understand.

Reply:
right, this shouldn't happen, but the server doesn't prevent it. I'll add a comment.